haystack/preview/utils/filters.py

```diff
@@ -292,3 +292,121 @@ def _list_conditions(conditions: Any) -> List[Any]:
     if isinstance(conditions, dict):
         return [{key: value} for key, value in conditions.items()]
     return [conditions]
+
+
+def convert(filters: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Convert a filter declared using the legacy style into the new style.
+    This is mostly meant to ease migration from Haystack 1.x to 2.x for developers
+    of Document Stores and Components that use filters.
+
+    This function doesn't verify if `filters` are declared using the legacy style.
+
+    Example usage:
+    ```python
+    legacy_filter = {
+        "$and": {
+            "type": {"$eq": "article"},
+            "date": {"$gte": "2015-01-01", "$lt": "2021-01-01"},
+            "rating": {"$gte": 3},
+            "$or": {"genre": {"$in": ["economy", "politics"]}, "publisher": {"$eq": "nytimes"}},
+        }
+    }
+    assert convert(legacy_filter) == {
+        "operator": "AND",
+        "conditions": [
+            {"field": "type", "operator": "==", "value": "article"},
+            {"field": "date", "operator": ">=", "value": "2015-01-01"},
+            {"field": "date", "operator": "<", "value": "2021-01-01"},
+            {"field": "rating", "operator": ">=", "value": 3},
+            {
+                "operator": "OR",
+                "conditions": [
+                    {"field": "genre", "operator": "in", "value": ["economy", "politics"]},
+                    {"field": "publisher", "operator": "==", "value": "nytimes"},
+                ],
+            },
+        ],
+    }
+    ```
+    """
+    converted = _internal_convert(filters)
+    if "conditions" not in converted:
```

Comment:
Are we sure this happens only in case of the simple filters you describe below? I would have handled this in the _internal_convert function (or somewhere else where it makes sense) and leave this check for conditions that really didn't get converter, to raise an exception.